satellite/metainfo/loop.go

```
@@ -0,0 +1,231 @@
+	return &Loop{
+		metainfo: metainfo,
+		config:   config,
+		join:     make(chan *observerContext),
+		done:     make(chan struct{}),
+	}
+}
+
+// Join will join the looper for one full cycle until completion and then returns.
+// On ctx cancel the observer will return without completely finishing.
+// Only on full complete iteration it will return nil.
+// Safe to be called concurrently.
+func (loop *Loop) Join(ctx context.Context, observer Observer) (err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	obsContext := &observerContext{
+		Observer: observer,
+		ctx:      ctx,
+		done:     make(chan error),
+	}
+
+	select {
+	case loop.join <- obsContext:
+	case <-ctx.Done():
+		return ctx.Err()
+	case <-loop.done:
+		return LoopClosedError
+	}
+
+	return obsContext.Wait()
+}
+
+// Run starts the looping service.
+// It can only be called once, otherwise a panic will occur.
+func (loop *Loop) Run(ctx context.Context) (err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	defer close(loop.done)
+
+	for {
+		err := loop.runOnce(ctx)
+		if err != nil {
+			return err
+		}
+	}
+}
+
+// runOnce goes through metainfo one time and sends information to observers.
+func (loop *Loop) runOnce(ctx context.Context) (err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	var observers []*observerContext
+	defer func() {
+		for _, observer := range observers {
+			observer.Finish()
+		}
+	}()
+
+	// wait for the first observer, or exit because context is canceled
+	select {
+	case observer := <-loop.join:
+		observers = append(observers, observer)
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+
+	// after the first observer is found, set timer for CoalesceDuration and add any observers that try to join before the timer is up
+	timer := time.NewTimer(loop.config.CoalesceDuration)
+waitformore:
+	for {
+		select {
+		case observer := <-loop.join:
+			observers = append(observers, observer)
+		case <-timer.C:
+			break waitformore
+		case <-ctx.Done():
+			for _, observer := range observers {
```

When the final err handling is part of the defer above, then you can return the error directly here.